nfs/src/rpc_task.cpp

```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
+    }
+
+    if (error_code < 0) {
+        AZLogDebug("copy_to_cache return error={}", error_code);
+
+        assert(chunkvec.size() == 0);
+        reply_error(error_code);
+        return;
+    }
+
+    assert(chunkvec.size() > 0);
+
+    for (auto &chunk : chunkvec) {
+        WRITE3args args;
+        bool rpc_retry = false;
+
+        // Get the membuf which need to flush.
+        auto membuf = chunk.get_membuf();
+        membuf->set_locked();
+
+        assert(membuf->is_dirty());
+        assert(membuf->is_uptodate());
+
+        // Fill the Write RPC arguments.
+        args.file = client->get_nfs_inode_from_ino(file_ino)->get_fh();
+        args.offset = membuf->offset;
+        args.count  = membuf->length;
+        args.stable = FILE_SYNC;
+        args.data.data_len = membuf->length;
+        args.data.data_val = (char *) membuf->buffer;
+
+        // Allocate new rpc task.
+        auto flush_task = client->get_rpc_task_helper()->alloc_rpc_task();
+        flush_task->optype = FUSE_FLUSH;
+
+        struct write_flush_context *cb_context = new write_flush_context(chunk, flush_task, file_ino);
+
+        // Keep the inuse and set flag to flushing.
+        membuf->set_flushing();
+
+        do 
+        {
+            if(rpc_nfs3_write_task(flush_task->get_rpc_ctx(), write_flush_callback, &args, cb_context) == NULL)
+            {
+                /*
+                * This call fails due to internal issues like OOM etc
+                * and not due to an actual error, hence retry.
+                */
+                rpc_retry = true;
+            }
+        } while (rpc_retry);
+
+       // AZLogDebug("membuf flushing offset {}, Length {}", membuf->offset, membuf->length);
+    }
+
+    // Send reply to original request.
+    reply_write(length);
+} 
+
+void rpc_task::run_flush()
+{
+    fuse_ino_t file_ino = rpc_api.write_task.get_ino();
+    auto nfs_inode = get_client()->get_nfs_inode_from_ino(file_ino);
+    auto filecache_handle = nfs_inode->filecache_handle;
+
+    assert(filecache_handle != nullptr);
+
+    // Get the dirty bytes_chunk from the filecache handle.
+    auto chunk_vec = filecache_handle->get_dirty_bc();
+
+    for (auto &chunk: chunk_vec)
+    {        
+        auto membuf = chunk.get_membuf();
+        assert(membuf != nullptr);
+
+        membuf->set_locked();
+        assert(membuf->is_inuse());
+
+        /*
+         * As of now, whenever write comes we get byte_chunk for corresponding
+         * offset and length. Copy the buffer from fuse to that, mark the buffer dirty.
+         * After that we issue write rpc to blob and complete the write command.
+         * Ideally all the chunks marked dirty in chunkmap, already issued to blob.
+         * We just need to wait for them to complete. That's what flush will do.
+         *
+         * Note: There is race condition to that, where we marked the buffer dirty, before
+         *       flushing it to blob, flush command come.
+         *
```

Comment:
we should not have to worry about these.
I suggest write a function called sync_membuf(). It takes a membuf parameter and ensures it's sync'ed to the backing blob. Right now you can write it here, later we can decide if that moves to the membuf class itself.
Once you have this function, then write()/flush()/sync() just call this function on all dirty membufs. write() gets dirty membufs from copy_to_cache() and flush/sync will get it using get_dirty_bc().
Then we don't need to worry about flush and write racing. We just need to make sure a dirty membuf is sync'ed.
Whoever gets to it firsy, syncs it, the other one doesn't do anything.

Reply:
Done